### PR TITLE
Interpreter: fix multidispatch with captured block

### LINKED ISSUE #12217: Interpreter: Nil assertion failed (NilAssertionError) from `spider-gazelle/promise`
Extracted from https://github.com/crystal-lang/crystal/issues/12214#issuecomment-1178377119

The Nil assertion error comes from shard [promise](https://github.com/spider-gazelle/promise).

Example code:

Reduced:
```crystal
class A
  def then(&callback)
    puts "from A"
    callback.call
  end
end

class B < A
  def then(&callback)
    puts "from B"
    callback.call
  end
end

foo = A.new || B.new

foo.then do
  puts "hello"
end

```
Error:
```
$ cr i then.cr 
Using compiled compiler at /home/chao/git/personal/crystal/.build/crystal
Nil assertion failed (NilAssertionError)
  from /home/chao/git/personal/crystal/src/nil.cr:108:5 in 'not_nil!'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/bindings.cr:13:7 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/main_visitor.cr:1066:7 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'create_def_uncached'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/multidispatch.cr:44:15 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:551:7 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/bindings.cr:17:7 in 'compile_def:closure_owner'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:2963:5 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:551:7 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:3204:11 in 'compile'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/interpreter.cr:220:5 in 'interpret'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/repl.cr:146:5 in 'interpret_and_exit_on_error'
  from /home/chao/git/personal/crystal/src/gc/boehm.cr:129:5 in 'repl'
  from /home/chao/git/personal/crystal/src/compiler/crystal/command.cr:100:7 in '__crystal_main'
  from /home/chao/git/personal/crystal/src/crystal/main.cr:115:5 in 'main'
  from /lib/x86_64-linux-gnu/libc.so.6 in '??'
  from /lib/x86_64-linux-gnu/libc.so.6 in '__libc_start_main'
  from /home/chao/git/personal/crystal/.build/crystal in '_start'
  from ???
```

Original:
```crystal
require "./src/promise"

LOG  = [] of Symbol
WAIT = Channel(Bool).new
promise = Promise.new(Symbol)
promise.then { |value| LOG << value; WAIT.send(true) }
```

Interpreter gives below error:
```
Nil assertion failed (NilAssertionError)
  from /home/chao/git/personal/crystal/src/nil.cr:108:5 in 'not_nil!'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/bindings.cr:13:7 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/main_visitor.cr:1066:7 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'create_def_uncached'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/multidispatch.cr:44:15 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:551:7 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/bindings.cr:17:7 in 'compile_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1969:5 in 'create_compiled_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1768:22 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:551:7 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/bindings.cr:17:7 in 'compile_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1969:5 in 'create_compiled_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1768:22 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:551:7 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/bindings.cr:17:7 in 'compile_block'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:2039:7 in 'create_compiled_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1768:22 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:551:7 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/bindings.cr:17:7 in 'compile_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1969:5 in 'create_compiled_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1768:22 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:551:7 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/bindings.cr:17:7 in 'compile_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1969:5 in 'create_compiled_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1768:22 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:551:7 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/bindings.cr:17:7 in 'compile_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1969:5 in 'create_compiled_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1768:22 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:476:7 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/bindings.cr:17:7 in 'compile_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:2981:3 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:3011:5 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:551:7 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/bindings.cr:17:7 in 'compile_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1969:5 in 'create_compiled_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1768:22 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/bindings.cr:17:7 in 'compile_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1969:5 in 'create_compiled_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1768:22 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:551:7 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/bindings.cr:17:7 in 'compile_block'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:2039:7 in 'create_compiled_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1768:22 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:551:7 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/bindings.cr:17:7 in 'compile_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1969:5 in 'create_compiled_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1768:22 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/bindings.cr:17:7 in 'compile_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1969:5 in 'create_compiled_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1768:22 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:433:7 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:551:7 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/bindings.cr:17:7 in 'compile_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:2981:3 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:3011:5 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:3012:5 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:551:7 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/bindings.cr:17:7 in 'compile_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1969:5 in 'create_compiled_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1768:22 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:3012:5 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:551:7 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/bindings.cr:17:7 in 'compile_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1969:5 in 'create_compiled_def'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:1768:22 in 'visit'
  from /home/chao/git/personal/crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:551:7 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/semantic/bindings.cr:17:7 in 'compile_def:closure_owner'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:2963:5 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:551:7 in 'accept'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/compiler.cr:3204:11 in 'compile'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/interpreter.cr:220:5 in 'interpret'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/repl.cr:146:5 in 'interpret_and_exit_on_error'
  from /home/chao/git/personal/crystal/src/gc/boehm.cr:129:5 in 'repl'
  from /home/chao/git/personal/crystal/src/compiler/crystal/command.cr:100:7 in '__crystal_main'
  from /home/chao/git/personal/crystal/src/crystal/main.cr:115:5 in 'main'
  from /lib/x86_64-linux-gnu/libc.so.6 in '??'
  from /lib/x86_64-linux-gnu/libc.so.6 in '__libc_start_main'
  from /home/chao/git/personal/crystal/.build/crystal in '_start'
```

### PR DESCRIPTION
**Pull request by asterite**
Fixes #12217

Note: this wasn't implemented at all, that's why it wasn't working.

### CODE REVIEW

File: spec/compiler/interpreter/multidispatch_spec.cr
Comment (straight-shoota) [inline spec/compiler/interpreter/multidispatch_spec.cr:422]: Due to the laws of commutativity, this spec would probably pass even if multi-dispatch selected exactly the opposite target def (`b` for `a || b` and `a` for `b || a`).
So I'd suggest to make the results clearly identifyable, for example with this (and expected value `1131`).
```suggestion
        a_value * 100 + b_value
```

Comment (asterite) [reply spec/compiler/interpreter/multidispatch_spec.cr]: Great point! I used subtraction instead.

**Review by straight-shoota — APPROVED** [review]

**Comment by asterite** [comment]
Is it okay if I merge interpreter PRs shortly (a few hours) after they've been approved? When I work on the next fix I'd like to have all other fixes in case they are somehow related.

Just asking because I've seen the usual approach is to wait a day or more to merge things before they are approved.

**Comment by asterite** [comment]
I'll merge it but let me know if you don't want me to do it like that.

**Pull request merged**
